### PR TITLE
[CSL-227] Library overhaul

### PR DESCRIPTION
**Pull request by sblaurock**
Now with:
- Methods to retrieve search, browse, autocomplete and recommendation results (similar in implementation to client-javascript)
- Coverage reports
- Minimal dependencies (just `node-fetch`, `qs`, and `form-data`)
- Ability to supply required backend parameters with requests to SABR endpoints
- Full JSDoc documentation (published to Github Pages)
- Consistency / quality of life improvements
- 92% test coverage

Concerns / follow ups:
- The current tests are more comprehensive in detail around validating use cases of specific parameters. We may want to follow up with porting these over to increase test strength. The ones in this PR do cover basic red / green paths though and ensure our newly implemented methods are working as expected.

Screen Shot 2021-04-22 at 10 31 38 PM

Screen Shot 2021-04-22 at 10 32 09 PM

Screen Shot 2021-04-22 at 10 17 34 PM

### CODE REVIEW

File: src/modules/catalog.js [hunk 1]
```diff
@@ -0,0 +1,1572 @@
...
+      }
+
+      return helpers.throwHttpErrorFromResponse(new Error(), response);
+    });
+  }
+
+  /**
+   * Remove item from index
+   *
+   * @function addOrUpdateItem
```
Comment (jjl014) [inline src/modules/catalog.js:164]: `addOrUpdateItem` -> `removeItem`

File: src/modules/catalog.js [hunk 2]
```diff
@@ -0,0 +1,1572 @@
...
+   * @param {string} parameters.id - Item group ID
+   * @param {string} parameters.name - Item group name
+   * @param {string} [parameters.parent_id] - Item group parent ID
+   * @returns {Promise}
+   * @see https://docs.constructor.io/rest_api/item_groups
+   */
+  addItemGroup(parameters = {}) {
+    let requestUrl;
+    const fetch = (this.options && this.options.fetch) || nodeFetch;
+    const { id, ...rest } = parameters;
```
Comment (jjl014) [inline src/modules/catalog.js:469]: Ooo, very neat! 😄

File: src/modules/catalog.js [hunk 3]
```diff
@@ -0,0 +1,1572 @@
...
+      if (response.ok) {
+        return response.json();
+      }
+
+      return helpers.throwHttpErrorFromResponse(new Error(), response);
+    }).then(json => json);
+  }
+
+  /**
+   * Remove all one way synonym
```
Comment (jjl014) [inline src/modules/catalog.js:860]: Description and @function detail below might need to be updated?

File: spec/src/modules/catalog.js [hunk 4]
```diff
@@ -0,0 +1,2625 @@
...
+        const itemGroupBuffer = fs.readFileSync(filePath);
+        const data = {
+          item_groups: itemGroupBuffer,
+          section: 'Products',
+        };
+
+        catalog.replaceCatalog(data).then(done);
+      });
+
+      it('Should replace a catalog of item groups using streams', (done) => {
```
Comment (jjl014) [inline spec/src/modules/catalog.js:2398]: Looks like this is a duplicate test, it exists on line 2366.

File: spec/src/modules/catalog.js [hunk 5]
```diff
@@ -0,0 +1,2625 @@
...
+        const itemGroupBuffer = fs.readFileSync(filePath);
+        const data = {
+          item_groups: itemGroupBuffer,
+          section: 'Products',
+        };
+
+        catalog.updateCatalog(data).then(done);
+      });
+
+      it('Should update a catalog of item groups using streams', (done) => {
```
Comment (jjl014) [inline spec/src/modules/catalog.js:2564]: Duplicate test, already exists on line 2532

**Review by jjl014 — CHANGES_REQUESTED** [review]
Nice, amazing amazing work on this! Totally killed it! 😄 

I left a couple comments, but otherwise looks great.

One other comment I had was around catalog upload. It looks like uploading variations without a matching top level item in the items.csv file causes the ingestion to fail. It looks like items.csv is required if you want to upload variations? 🤔


Screen Shot 2021-04-27 at 5 43 33 PM

**Review by crgee1 — COMMENTED** [review]
You and @esezen did a great job. Other than Jimmy's comments, everything looks good

File: spec/src/modules/catalog.js (hunk 4 above)
Comment (esezen) [reply spec/src/modules/catalog.js:2398]: I was trying to make it look like I wrote more tests. Caught

**Comment by sblaurock** [comment]
> Nice, amazing amazing work on this! Totally killed it! 😄
> 
> I left a couple comments, but otherwise looks great.
> 
> One other comment I had was around catalog upload. It looks like uploading variations without a matching top level item in the items.csv file causes the ingestion to fail. It looks like items.csv is required if you want to upload variations? 🤔
> 
> Screen Shot 2021-04-27 at 5 43 33 PM

Interesting! Is this something with our system or this PR specifically? I guess it does make sense that you can't upload variations without items unless it's a delta and they already exist in the system

**Review by jjl014 — APPROVED** [review]
🚀🚢

**Comment by jjl014** [comment]
> Interesting! Is this something with our system or this PR specifically? I guess it does make sense that you can't upload variations without items unless it's a delta and they already exist in the system

It's something with the system, not the PR. Was just wondering if we should also add a restriction on the client side.

**Pull request merged**